File: ADT/kruskals_algo/kruskal's_algo.py


# Adj -> {{u1 ,v1, w1}, {u2, v2, w2}...}
# Sort vec basis on weight (ascending)
# Start processing from smallest edges

# Component detection by DSU
# If they are already connected, don't do anything, if not connected, connect the smalles edges
# By using DSU find parent, if parent are not same that means they are in different component so union them to make them locate in same components
# If belong to same components, somehow they connected so need to connect again


class Solution:
    def __init__(self):
        # parent[i] will store the parent of node i
        self.parent = []

        # rank[i] is used to keep tree depth small
        self.rank = []

    def find(self, x):
        """
        Finds the ultimate parent (root) of x
        Uses path compression for optimization
        """
        # If x is its own parent, it is the root
        if self.parent[x] == x:
            return x

        # Recursively find parent and compress path
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        """
        Unites the sets containing x and y
        Uses union by rank
        """
        # Find root parents
        x_parent = self.find(x)
        y_parent = self.find(y)

        # If already in same set, do nothing
        if x_parent == y_parent:
            return

        # Attach smaller rank tree under larger rank tree
        if self.rank[x_parent] > self.rank[y_parent]:
            self.parent[y_parent] = x_parent
        elif self.rank[x_parent] < self.rank[y_parent]:
            self.parent[x_parent] = y_parent
        else:
            # If ranks are equal, choose one and increase its rank
            self.parent[x_parent] = y_parent
            self.rank[y_parent] += 1

    def kruskal(self, edges):
        """
        Applies Kruskal's algorithm on sorted edge list
        """
        mst_weight = 0

        # Traverse edges in increasing order of weight
        for u, v, wt in edges:
            # If u and v belong to different components
            if self.find(u) != self.find(v):
                # Include this edge in MST
                self.union(u, v)
                mst_weight += wt

        return mst_weight

    def spanningTree(self, V, adj):
        """
        Main function called by driver code
        """

        # Initialize DSU
        self.parent = [i for i in range(V)]
        self.rank = [0] * V

        edges = []

        # Convert adjacency list to edge list
        for u in range(V):
            for v, wt in adj[u]:
                # Avoid duplicate edges (undirected graph)
                if u < v:
                    edges.append((u, v, wt))

        # Sort edges by weight
        edges.sort(key=lambda x: x[2])

        # Run Kruskal algorithm
        return self.kruskal(edges)

File: ADT/kruskals_algo/kruskal's_algo.js
class Solution {
    constructor() {
        // parent[i] stores the parent of node i
        this.parent = [];

        // rank[i] helps keep the tree shallow
        this.rank = [];
    }

    find(x) {
        /**
         * Finds the root of x
         * Uses path compression
         */
        if (this.parent[x] === x) {
            return x;
        }

        // Compress path
        this.parent[x] = this.find(this.parent[x]);
        return this.parent[x];
    }

    union(x, y) {
        /**
         * Unites the sets of x and y using rank
         */
        let xParent = this.find(x);
        let yParent = this.find(y);

        // Already in same set
        if (xParent === yParent) return;

        // Union by rank
        if (this.rank[xParent] > this.rank[yParent]) {
            this.parent[yParent] = xParent;
        } else if (this.rank[xParent] < this.rank[yParent]) {
            this.parent[xParent] = yParent;
        } else {
            this.parent[xParent] = yParent;
            this.rank[yParent]++;
        }
    }

    kruskal(edges) {
        /**
         * Runs Kruskal algorithm on sorted edges
         */
        let mstWeight = 0;

        for (let [u, v, wt] of edges) {
            // Check if adding edge creates a cycle
            if (this.find(u) !== this.find(v)) {
                this.union(u, v);
                mstWeight += wt;
            }
        }

        return mstWeight;
    }

    spanningTree(V, adj) {
        /**
         * Main MST function
         */

        // Initialize DSU arrays
        this.parent = new Array(V);
        this.rank = new Array(V).fill(0);

        for (let i = 0; i < V; i++) {
            this.parent[i] = i;
        }

        let edges = [];

        // Convert adjacency list to edge list
        for (let u = 0; u < V; u++) {
            for (let [v, wt] of adj[u]) {
                // Avoid duplicate edges
                if (u < v) {
                    edges.push([u, v, wt]);
                }
            }
        }

        // Sort edges by weight
        edges.sort((a, b) => a[2] - b[2]);

        // Apply Kruskal
        return this.kruskal(edges);
    }
}
